### ISSUE TITLE
[Feature request] Support Azure OpenAI API

### ISSUE THREAD

**Issue by tuntunhuang** [brianpetro/obsidian-smart-connections#382]
Hi Brain,
Thank you very much for developing this amazing plugin. Would you update it to support the Azure OpenAI API?

**Comment by carllx** [comment]
I completely agree. Allowing users to customize the API base URL, rather than hard-coding it to "https://api.openai.com/v1/chat/completions", would be extremely helpful in the current situation. Students living outside the service area of OpenAI would benefit from this.

**Comment by u3588064** [comment]
+1. Can't wait to use Azure API.

**Comment by gaojunyang666** [comment]
I completely agree. Allowing users to customize the API base URL, rather than hard-coding it to "https://api.openai.com/v1/chat/completions", would be extremely helpful in the current situation. Students living outside the service area of OpenAI would benefit from this.

Yes, I think this is very important. I have built an API myself, and I hope to change the interface to my own interface

**Comment by brianpetro** [comment]
@gaojunyang666 thanks for the feedback 

This is already partially implemented in `v2.1` for use with localhost APIs. A custom non-localhost endpoint, assuming that it uses the OpenAI API format, can be implemented rather quickly, but so far this hasn't been requested by anyone helping with beta testing the early release.

For non-OpenAI formats, there will be the ability to contribute additional formats to the module I created for implementing these models. This module will be released as an open-source NPM module very soon.

🌴

**Comment by ChrisRomp** [comment]
This is *almost* there using `Custom API (OpenAI format)` except that it sends the `api-key` header as `Authorization: Bearer xxxx`.  If we could override the auth header, I could make this work with Azure OpenAI.

**Comment by ChrisRomp** [comment]
I can (eventually) take a crack at it. I just want to test this in my Obsidian workflow today. If it works well for me then I'll try to make some time for this.

But the gist for anyone else is it should exactly like an OpenAI endpoint, except you need to provide a couple of params in the URL (see [docs](https://learn.microsoft.com/en-us/azure/ai-services/openai/reference#chat-completions)). E.g.:

```http
POST https://{your-resource-name}.openai.azure.com/openai/deployments/{deployment-id}/chat/completions?api-version={api-version}
```

For anyone wanting to use this on Azure today, I've put Azure API Management in front of AOAI, and I'm using policy to compare a key to a named value in APIM. Then I use Azure RBAC auth to the AOAI service. You could optionally extract the Bearer value and pass that into an `api-key` header to AOAI instead of storing it as a secret in APIM. Disable requiring a subscription key on the API as well.

```xml
<policies>
    <inbound>
        <base />
        <check-header name="Authorization" failed-check-httpcode="401" failed-check-error-message="Unauthorized">
            <value>Bearer {{aoai-custom-key}}</value>
        </check-header>
        <authentication-managed-identity resource="https://cognitiveservices.azure.com" output-token-variable-name="msi-access-token" ignore-error="false" />
        <set-header name="Authorization" exists-action="override">
            <value>@("Bearer " + (string)context.Variables["msi-access-token"])</value>
        </set-header>
    </inbound>
    <backend>
        <base />
    </backend>
    <outbound>
        <base />
    </outbound>
    <on-error>
        <base />
    </on-error>
</policies>
```

You can import the Swagger spec for AOAI into APIM from the Azure docs: https://learn.microsoft.com/en-us/azure/ai-services/openai/reference

**Comment by ChrisRomp** [comment]
@brianpetro I could also give you short term access to an Azure OpenAI instance.

**Comment by infinitelyloopy-bt** [comment]
> @gaojunyang666 thanks for the feedback
> 
> This is already partially implemented in `v2.1` for use with localhost APIs. A custom non-localhost endpoint, assuming that it uses the OpenAI API format, can be implemented rather quickly, but so far this hasn't been requested by anyone helping with beta testing the early release.
> 
> For non-OpenAI formats, there will be the ability to contribute additional formats to the module I created for implementing these models. This module will be released as an open-source NPM module very soon.
> 
> 🌴

Hey Brian,

Any news on allowing us to enter our own API keys and URL for Azure OpenaI?

**Comment by ujikol** [comment]
With 2.3.49 it is possible to specify the correct URI.
The remaining issue seems to be that the API is provided as Auth header instead of a api-key parameter.
Any simple fix for that, which does not need any changes on Azure side? (e.g. available local proxy to patch the requests)
Of course a real fix in a new version would be great. Thx :)

**Comment by brianpetro** [comment]
@ujikol thanks for following up. I'm going to try to fit this in shortly as I have a couple other updates to make in the chat model adapters 🌴

**Comment by ujikol** [comment]
With 2.3.52 I now get this error:
plugin:smart-connections:14705 Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'startsWith')

The call of deployments brings 404. The endpoint is called with the API version as parameter, which does not make sense. Is this endpoint even available for Azure OpenAI?

The setting I made work perfectly fine with the Obsidian Copilot plugin.

**Comment by ujikol** [comment]
![Image](https://github.com/user-attachments/assets/e9a01453-8e4c-432c-a7a2-3809c78f5271)

While I cannot give you access to the Azure tenant, maybe you want to give me some curl requests (in Windows format) and the expected results to test?

**Comment by brianpetro** [comment]
@ujikol that should be what I need. I'm looking into it now 🌴

**Comment by brianpetro** [comment]
@ujikol the bug in the screenshot is fixed in `v2.3.53`.

Let me know if that gets things running. Or if another issue pops up that needs to be addressed.

Thanks for your help 🌴

**Comment by ujikol** [comment]
It works now. 😃 
Thank you! 💯

**Comment by ChrisRomp** [comment]
I'm not able to configure the chat settings (Smart Threads Settings) w/ my AOAI info. I can see it's making an API call to `https://my-resource.openai.azure.com/openai/deployments` but that's not a valid REST endpoint per the [docs](https://learn.microsoft.com/en-us/azure/ai-services/openai/reference).  I also noticed it's including an `Authorization` header with a value of `Bearer [api-key]` when it should be using a header `api-key` with a value of `[api-key]`, e.g. `curl -H "api-key: asdf1234" https://my-resource.openai.azure.com/openai/....`

The result of all of this is I'm unable to select a chat model, as it says "No deployments found."

![Image](https://github.com/user-attachments/assets/830d5d24-a02e-412d-8364-db9f0677bd10)

**Comment by yvesgoeleven** [comment]
@ChrisRomp I've found a workaround to get smart chat working with azure openai.

The settings themselves are stored in .obsidian/.smart-env/smart_threads/smart_env.json

"azure": {
        "model_key": "your-model",
        "azure_resource_name": "your-resource",
        "azure_deployment_name": "your-deployment",
        "azure_api_version": "2025-01-01-preview",
        "api_key": "your-key",
        "models": {
          "_": {
            "id": "your-model"
          }
        }
      },

You can fill in the details directly.

To workaround the apikey header, navigate to

.obsidian\plugins\smart-connections\main.js

Find the class SmartChatModelAzureAdapter and remove the default for `api_key_header: "api-key",`, without it the api proxy will fall back to the Authorization header.

I don't have a fix yet for model selection in the settings pane though, it seems the old api got discontinued but it's replacement needs more information than is currenly available in the thread settings.

The replacement is

https://management.azure.com/subscriptions/{subscription-id}/resourceGroups/{resource-group}/providers/Microsoft.CognitiveServices/accounts/{resource-id}/deployments?api-version=2025-01-01-preview
